### DISCUSSION TITLE
Lesson 4 -  Beginner to Expert Python Tutorial - 3:56:40 - Having alot of trouble using python deploy.py

### DISCUSSION THREAD

**Discussion by richwealthyandsuccessful** [smartcontractkit/full-blockchain-solidity-course-py#1582] [Q&A]
At present my code is...

```
from solcx import compile_standard, install_solc
import json
from web3 import Web3

with open("./SimpleStorage.sol", "r") as file:
    simple_storage_file = file.read()

print("Installing...")
install_solc("0.6.0")

# Solidity source code
compiled_sol = compile_standard(
    {
        "language": "Solidity",
        "sources": {"SimpleStorage.sol": {"content": simple_storage_file}},
        "settings": {
            "outputSelection": {
                "*": {
                    "*": ["abi", "metadata", "evm.bytecode", "evm.bytecode.sourceMap"]
                }
            }
        },
    },
    solc_version="0.6.0",
)
with open("compile_code.json", "w") as file:
    json.dump(compiled_sol, file)


# get bytecode
bytecode = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["evm"][
    "bytecode"
]["object"]

# get abi
abi = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["abi"]

# for conneting to ganache
w3 = Web3(Web3.HTTPProvider("http://127.0.0.1:7545"))
chain_id = 1337
my_address = "0x66487dD584Ed8bB02064f2d855659882a5772134"
private_key = "0xe02ab2b4fddde5ba4ec22fc971359983e2687b21cf30544d85ab2991080aec13"

# create the contract in python
SimpleStorage = w3.eth.contract(abi=abi, bytecode=bytecode)
print(SimpleStorage)

# get the latest transactions
nonce = w3.eth.getTransactionCount(my_address)
print(nonce)
```
But when I use `python3 deploy.py` or `python deploy.py` all I'm receiving is errors like this....

```
Traceback (most recent call last):
  File "/home/marcusdahibbert/demos/web3_py_simple_storage/web3_py_simple_storage/deploy.py", line 9, in <module>
    from dotenv import load_dotenv
ModuleNotFoundError: No module named 'dotenv'
```
Not sure where I'm going wrong here? Could anybody help, so I can continue with this tutorial.

For some more information, I'm using a Chromebook, Chrome OS, VSCode and I've downloaded the dependencies, but currently using Ganache's extension from VSCode.

**Comment by cromewar** [comment]
Hello @richwealthyandsuccessful 
`dotenv` is for charging path variables from a file called `.env` it says you are importing that library at line 9, so please check if you have that extra import on `deploy.py` , also just install the package git with pip using the official command:
https://pypi.org/project/python-dotenv/

**Reply by richwealthyandsuccessful** [reply]
Thank you cromewar, I will take onboard your advice above and let you know how I get on.

**Reply by cromewar** [reply]
Please let us know if this information was helpful for you @richwealthyandsuccessful

**Reply by richwealthyandsuccessful** [reply]
Hi cromewar, so I followed your advice. I don't think I had the extra import contained within deploy.py so I just added it like this:

```
_from solcx import compile_standard, install_solc
import json
from web3 import Web3

with open("./SimpleStorage.sol", "r") as file:
    simple_storage_file = file.read()

from dotenv import load_dotenv

load_dotenv()  # take environment variables from .env.
```
I followed the link you provided and completed the instructions: `pip install python-dotenv` then attempted to deploy my contract again with `python3 deploy.py` but again we we're hit with an error.

```
Installing...
Deploying Contract!
Traceback (most recent call last):
  File "/home/marcusdahibbert/demos/web3_py_simple_storage/web3_py_simple_storage/deploy.py", line 76, in <module>
    tx_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
  File "/home/marcusdahibbert/.local/lib/python3.9/site-packages/web3/eth.py", line 831, in send_raw_transaction
    return self._send_raw_transaction(transaction)
  File "/home/marcusdahibbert/.local/lib/python3.9/site-packages/web3/module.py", line 57, in caller
    result = w3.manager.request_blocking(method_str,
  File "/home/marcusdahibbert/.local/lib/python3.9/site-packages/web3/manager.py", line 198, in request_blocking
    return self.formatted_response(response,
  File "/home/marcusdahibbert/.local/lib/python3.9/site-packages/web3/manager.py", line 171, in formatted_response
    raise ValueError(response["error"])
ValueError: {'message': 'insufficient funds for gas * price + value', 'stack': 'Error: insufficient funds for gas * price + value\n    at TransactionPool.prepareTransaction (/usr/lib/node_modules/ganache/dist/node/1.js:2:133762)', 'code': -32003}
```
Are we missing something? Because I believe the account had sufficient funds.

**Comment by richwealthyandsuccessful** [comment]
Just attempted again with a new account and new private key, but the same error appeared.

**Reply by cromewar** [reply]
how much ether do you have on that account? because based on the error message it seems you don't have enough founds to fulfill the transaction. Also you can set manually de gas limit on the transaction, that should help.

**Comment by richwealthyandsuccessful** [comment]
It says 1000 ETH in all the accounts which I'm attempting to use for deploying the contract, I'm using ganache from the terminal because I'm using the VSCode Ganache Extension, I was unable to get ganache to open with the visual mnemonic. 

The gas price is set to: 2000000000, which I'm assuming is "wei" rather than "eth"

I'm not sure whats wrong here in that case? I'm not entirely sure I can amend the gas price through the terminal.

**Reply by cromewar** [reply]
is there a VSCode Ganache Extension? never seen it, may be this could be the problem.
Please install node lts version 16, and install ganache trough npm and on a second terminal keep running the ganache server.
Did this help?

**Reply by richwealthyandsuccessful** [reply]
I did attempt this before but failed, but I'll try to do it again and hopefully this will solve the issue. The VScode gansche extension is relatively new I think, but i thought it might be another option seen as i couldn't make progress installing ganache:

https://trufflesuite.com/blog/build-on-web3-with-truffle-vs-code-extension/)

I'll let you know how this goes, thanks again mate.

**Reply by cromewar** [reply]
Mmm I would says the most probably scenario is the extension is not working properly could you try the the traditional approach and install ganache trough npm?

**Comment by richwealthyandsuccessful** [comment]
Okay, I ran `npm install ganache --global` this what I'm receiving:

```
npm WARN checkPermissions Missing write access to /usr/lib/node_modules/ganache
npm WARN checkPermissions Missing write access to /usr/lib/node_modules
npm ERR! code EACCES
npm ERR! syscall access
npm ERR! path /usr/lib/node_modules/ganache
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, access '/usr/lib/node_modules/ganache'
npm ERR!  [Error: EACCES: permission denied, access '/usr/lib/node_modules/ganache'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'access',
npm ERR!   path: '/usr/lib/node_modules/ganache'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/marcusdahibbert/.npm/_logs/2022-07-21T04_41_44_128Z-debug.log
```
I spoke to Trufflesuite customer support and they said that because I'm using a Chromebook with ChromeOS/ Linux Beta there might be comparability issues which is causing errors.

Does this mean I need a mac to proceed?

**Reply by cromewar** [reply]
Most likely is because the way you are installing npm which requires some low level permissions, is preferable to install npm trough [NVM](https://github.com/nvm-sh/nvm) which does not requires those privilegies, give it a try.

**Reply by akhilparakka** [reply]
I had the same issue. After starting terminal in VS Code, I had to downgrade node to version 16.4.
$nvm install 16.4

**Reply by richwealthyandsuccessful** [reply]
Okay, so I'll try both of your instructions and I'll report back. Thanks again

**Comment by richwealthyandsuccessful** [comment]
Okay, so since running: `nvm install 16.4` I've managed to get pass the last error, 

I then ran the same command to attempt to download ganache: `npm install ganache --global`

This seems to have worked now, as I'm getting the following: `ganache v7.3.2 (@ganache/cli: 0.4.2, @ganache/core: 0.4.2)`

Okay, so can I now proceed to deploy my contract? Because previously I had trouble actually getting the Ganache Workspace to open up (the quick start home screen). It wouldn't allow me to double click the file like you would typically on a windows or mac.

**Reply by cromewar** [reply]
Well brownie should execute your npm ganache automatically on the console, so this should work fine.

**Reply by richwealthyandsuccessful** [reply]
okay great, well I'll continue and let you know how it all went. Thanks for your help @cromewar
